rust/frontend/src/function.rs

```diff
@@ -0,0 +1,476 @@
+        self.clone()(())
+    }
+}
+
+impl<'a> FnOnce<((),)> for Builder<'a> {
+    type Output = Result<TVMRetValue<'a>>;
+    extern "rust-call" fn call_once(self, _: ((),)) -> Self::Output {
+        if self.func.is_none() {
+            return Err(Error::NoFunction);
+        }
+        let mut ret_val = unsafe { mem::uninitialized::<ts::TVMValue>() };
+        let mut ret_type_code = 0 as c_int;
+        if self.arg_buf.is_some() {
+            let arg_buf = self.arg_buf.unwrap();
+            let mut num_args = arg_buf.len();
+            let mut values = arg_buf
+                .iter()
+                .map(|tav| tav.value.inner)
+                .collect::<Vec<ts::TVMValue>>();
+            let mut tcodes = arg_buf
+                .iter()
+                .map(|tav| tav.type_code as c_int)
+                .collect::<Vec<_>>();
+            if self.ret_buf.is_some() {
+                num_args = num_args + 1;
+                ret_val = *self.ret_buf.clone().unwrap()[0].value;
+                ret_type_code = self.ret_buf.clone().unwrap()[0].type_code as c_int;
+                values.append(&mut vec![ret_val]);
+                tcodes.append(&mut vec![ret_type_code]);
+            }
+            values.truncate(num_args);
+            tcodes.truncate(num_args);
+            check_call!(ts::TVMFuncCall(
+                self.func.unwrap().handle,
+                values.as_mut_ptr(),
+                tcodes.as_mut_ptr(),
+                num_args as c_int,
+                &mut ret_val as *mut _,
+                &mut ret_type_code as *mut _
+            ));
+        } else {
+            check_call!(ts::TVMFuncCall(
+                self.func.unwrap().handle,
+                ptr::null_mut(),
+                ptr::null_mut(),
+                0 as c_int,
+                &mut ret_val as *mut _,
+                &mut ret_type_code as *mut _
+            ));
+        }
+        let ret = TVMRetValue::new(
+            TVMValue::new(ValueKind::Return, ret_val),
+            ret_type_code.into(),
+        );
+        Ok(ret)
+    }
+}
+
+/// Converts a [`Function`] to builder. Currently, this is the best way to work with
+/// TVM functions.
+impl<'a> From<Function> for Builder<'a> {
+    fn from(func: Function) -> Self {
+        Builder::new(Some(func), None, None)
+    }
+}
+
+/// Converts a mutable reference of a [`Module`] to builder.
+impl<'a: 'b, 'b> From<&'b mut Module> for Builder<'a> {
+    fn from(module: &mut Module) -> Self {
+        Builder::new(module.entry.take(), None, None)
+    }
+}
+
+unsafe extern "C" fn tvm_callback(
+    args: *mut ts::TVMValue,
+    type_codes: *mut c_int,
+    num_args: c_int,
+    ret: ts::TVMRetValueHandle,
+    fhandle: *mut c_void,
+) -> c_int {
+    let len = num_args as usize;
+    let args_list = unsafe { slice::from_raw_parts_mut(args, len).to_vec() };
+    let type_codes_list = unsafe { slice::from_raw_parts_mut(type_codes, len).to_vec() };
```

Comment:
you don't need to `to_vec` either of these since nothing actually takes ownership of either `args_list` or `type_codes_list`.